### DISCUSSION TITLE
How to obtain passband luminosity ratio？

### DISCUSSION THREAD

**Discussion by alan0986** [phoebe-project/phoebe2#662] [General]
Hello,
I need to obtain the luminosity ratio of the two components when I run mcmc. However, in the 'component-coupled' mode, the secondary component luminosity is not visible. I try to get the luminosity ratio with the following custom constraints, and transfer the luminosity uncertainties of the primary component to the luminosity ratio after the mcmc runs.

lumidata = b.compute_pblums()
pblumratio_def = phoebe.parameters.FloatParameter(qualifier='pblumratio',
                                                 default_unit=u.dimensionless_unscaled,
                                                 value=1, 
                                                 description='Passband Luminosity ratio', overwrite_all=True)
pblumratio_param, created = b.get_or_create('pblumratio',pblumratio_def, context='dataset', dataset='lcT')
lsh = pblumratio_param
rsh = np.float64(lumidata['pblum@secondary@lcT'])/b.get_parameter(qualifier='pblum', component='primary', dataset='lcT')
b.add_constraint(lsh,rsh)

However, when I run
b.uncertainties_from_distribution_collection(solution='emcee_sol', parameters=['pblumratio'], tex=True)

NotImplementedError: cannot convert object with physical_ type=power/radiant flux to solar units

Or is there any other way to solve this problem?
Thanks

**Comment by kecnry** [comment]
I think your units error is because you are casting the `pblum@secondary` to a float, which then results in setting the expression with non-unitless units.  In theory you would divide two parameters, but as you've mentioned, the secondary luminosity isn't a parameter when the mode is set to component-coupled.  Even if this didn't raise a units error, this would be using a _fixed_ value for the secondary luminosity while computing the luminosity ratio - is that what you are intending to do?

Since you're using component-coupled, both luminosities will likely be changing, and so you would need to call `b.compute_pblums` for each value of the primary's luminosity.  Unfortunately, PHOEBE doesn't yet support a nice way of doing this, but depending on your use case, you might find the discussions in #571 (essentially the same use-case but just to access the secondary luminosity instead of the ratio) or possibly #546 and #443 useful.  I'll also just link to this [discussion topic for possibly adding native support within phoebe](https://github.com/phoebe-project/phoebe2/discussions/407#discussioncomment-335433) if you have any thoughts you want to add there.

**Reply by alan0986** [reply]
Thank you for your answer. I tried to use # 571 method to get the distribution of 'pblum@secondary stars' in the 'component coupled' mode. However, However, Phoebe will prompt the following errors. Is it still the error caused by the unit? 

File C:\ProgramData\Anaconda3\envs\name_of_phoebe_environment\lib\site-packages\distl\distl.py:413, in is_list_of_dists_or_floats(value)
    410         value[i] = from_dict(dist)
    412 if not _all_in_types(value, (BaseUnivariateDistribution, BaseMultivariateSliceDistribution, float, int)):
--> 413     raise TypeError("must be a tuple or list of distributions (univariate or multivariate slices) or floats")
    415 return value

TypeError: must be a tuple or list of distributions (univariate or multivariate slices) or floats

Thanks.

**Reply by kecnry** [reply]
I don't believe units would cause this, can you post the snippet of code that resulted in this error?

**Reply by alan0986** [reply]
Ok，OK, I can get the distribution of 'pblum@primary stars' . There may be obvious mistakes in my code. I hope you can give me some suggestions. Thanks.

```
def compute_pblum_secondary(b, pblum_primary):
    b.set_value('pblum@primary@lcT@lc@dataset', value=pblum_primary)
    # b.set_value('teff@secondary@star@component', value=teff_secondary)  
    return b.compute_pblums()['pblum@secondary@lcT@lc@dataset']

lp1=b.get_parameter('pblum@primary@lcT@lc@dataset',  ).get_distribution('emcee_sol')
lp1
#lp2=b.get_parameter('teff@secondary@star@component', ).get_distribution('emcee_sol')
lp3 = distl.function(compute_pblum_secondary, args=(b, lp1), vectorized=False)
```

**Reply by kecnry** [reply]
I will need to try to reproduce this or write up a full working example, but that will take a little time.  If you manage to find the solution in the meantime, please let us know.  Just to clarify - what version of PHOEBE are you using?

**Reply by alan0986** [reply]
Ok, I am using PHOEBE 2.4. Thanks again for your help.

**Reply by kecnry** [reply]
The issue was that the function does not accept taking the bundle itself as an argument, so you just need to remove that - sorry for the confusion (I just wrote the original pseudocode from memory).  I now created a full [example for propagating distributions through luminosity coupling](http://phoebe-project.org/docs/latest/examples/distribution_pblum) that includes this syntax fix and hopefully explains in more detail.  You'll still need to modify this logic slightly if you want the ratio instead of the secondary luminosity (in addition to the specific distributions that you have sampled from MCMC).  Let me know if you have any suggestions on how to improve the example or can't figure out how to modify it for your situation.

**Reply by alan0986** [reply]
Thank you very much for your patience and help.
